src/infill/SierpinskiFill.cpp

```diff
@@ -0,0 +1,802 @@
+        {
+            SierpinskiTriangle* node = *it;
+            depth_ordered[node->depth].emplace_back(it);
+        }
+    }
+
+    for (std::list<std::list<SierpinskiTriangle*>::iterator>& depth_nodes : depth_ordered)
+        for (std::list<SierpinskiTriangle*>::iterator it : depth_nodes)
+        {
+            redistributeLeftoverErrors(it, std::next(it), false);
+        }
+}
+
+void SierpinskiFill::diffuseError()
+{
+    int pair_constrained_nodes = 0;
+    int constrained_nodes = 0;
+    int unconstrained_nodes = 0;
+    int subdivided_nodes = 0;
+    float error = 0;
+    for (std::list<SierpinskiTriangle*>::iterator it = sequence.begin(); it != sequence.end(); ++it)
+    {
+        SierpinskiTriangle* node = *it;
+        SierpinskiTriangle& triangle = *node;
+
+        float boundary = (triangle.realized_length + triangle.total_child_realized_length) * .5f;
+        
+        float nodal_value = ((use_errors_in_dithering)? node->getErroredValue() : node->requested_length);
+        
+        float boundary_error = nodal_value - boundary + error;
+
+        std::list<SierpinskiTriangle*>::iterator begin = it;
+        std::list<SierpinskiTriangle*>::iterator end = std::next(it);
+        if (
+            triangle.dir == SierpinskiTriangle::SierpinskiDirection::AC_TO_AB
+            && end != sequence.end()
+        )
+        {
+            pair_constrained_nodes++;
+            continue; // don't subdivide these two triangles just yet, wait till next iteration
+        }
+        if (triangle.dir == SierpinskiTriangle::SierpinskiDirection::AB_TO_BC
+            && begin != sequence.begin()
+        )
+        {
+            begin = std::prev(it);
+            assert((*begin)->depth == triangle.depth || isConstrainedBackward(it));
+        }
+
+
+        bool is_constrained = false;
+        for (auto nested_it = begin; nested_it != end; ++nested_it)
+        {
+            if (isConstrainedBackward(nested_it) || isConstrainedForward(nested_it))
+            {
+                is_constrained = true;
+                constrained_nodes++;
+                break;
+            }
+        }
+        if (!is_constrained)
+            unconstrained_nodes++;
+        if (!is_constrained
+            && boundary_error >= 0
+            && !node->children.empty()
+            )
+        {
+            subdivided_nodes++;
+            it = subdivide(begin, end, false);
+            if (dithering)
+                error += nodal_value - triangle.total_child_realized_length;
+        }
+        else
+        {
+            if (dithering)
+                error += nodal_value - triangle.realized_length;
+        }
+    }
+    std::cerr << "pair_constrained_nodes: " << pair_constrained_nodes << ", constrained_nodes: " << constrained_nodes << ", unconstrained_nodes: " << unconstrained_nodes << ", subdivided_nodes: " << subdivided_nodes << '\n';
```

Comment:
We should use the logger for this sort of thing.

Reply:
yes. fixed